src/core/utils/power_calibration.cpp

```diff
@@ -0,0 +1,223 @@
+    CalibratedPower *calibratedPowers;
+
+    VerifyOrExit(IsChannelValid(aChannel) && (aRawPowerSetting != nullptr), error = kErrorInvalidArgs);
+    VerifyOrExit((mChannel != aChannel) || mPowerUpdated);
+
+    chIndex             = aChannel - Radio::kChannelMin;
+    targetPower         = mTargetPowerTable[chIndex];
+    calibratedPowers    = &mCalibrationPowerTable[chIndex][0];
+    numCalibratedPowers = mNumCalibratedPowers[chIndex];
+
+    VerifyOrExit(targetPower != CalibratedPower::kInvalidPower, error = kErrorNotFound);
+    VerifyOrExit(numCalibratedPowers > 0, error = kErrorNotFound);
+
+    for (i = numCalibratedPowers - 1; (i >= 0) && (targetPower < calibratedPowers[i].GetActualPower()); i--)
+        ;
+
+    VerifyOrExit(i >= 0, error = kErrorNotFound);
+
+    mCalibratedPower = &calibratedPowers[i];
+
+    mChannel      = aChannel;
+    mPowerUpdated = false;
+
+exit:
+    if (error == kErrorNone)
+    {
+        error = mCalibratedPower->GetRawPowerSetting(aRawPowerSetting, aRawPowerSettingLength);
+    }
+
+    return error;
+}
+
+bool PowerCalibration::FindCalibratedPower(CalibratedPower *aCalibratedPowers,
+                                           uint8_t          aNumCalibratedPowers,
+                                           int16_t          aActualPower)
+{
+    bool ret = false;
+
+    for (uint8_t i = 0; i < aNumCalibratedPowers; i++)
+    {
+        VerifyOrExit(aCalibratedPowers[i].GetActualPower() != aActualPower, ret = true);
+    }
+
+exit:
+    return ret;
+}
+} // namespace Utils
+} // namespace ot
+
+using namespace ot;
+
+OT_TOOL_WEAK otError otPlatRadioAddCalibratedPower(otInstance *   aInstance,
+                                                   uint8_t        aChannel,
+                                                   int16_t        aActualPower,
+                                                   const uint8_t *aRawPowerSetting,
+                                                   uint16_t       aRawPowerSettingLength)
+{
+    return AsCoreType(aInstance).Get<Utils::PowerCalibration>().AddCalibratedPower(
+        aChannel, aActualPower, aRawPowerSetting, aRawPowerSettingLength);
+}
+
+OT_TOOL_WEAK otError otPlatRadioClearCalibratedPowers(otInstance *aInstance)
+{
+    AsCoreType(aInstance).Get<Utils::PowerCalibration>().ClearCalibratedPowers();
+    return OT_ERROR_NONE;
+}
+
+OT_TOOL_WEAK otError otPlatRadioSetChannelTargetPower(otInstance *aInstance, uint8_t aChannel, int16_t aTargetPower)
+{
+    return AsCoreType(aInstance).Get<Utils::PowerCalibration>().SetChannelTargetPower(aChannel, aTargetPower);
+}
+
+otError otPlatRadioGetRawPowerSetting(otInstance *aInstance,
+                                      uint8_t     aChannel,
+                                      uint8_t *   aRawPowerSetting,
+                                      uint16_t *  aRawPowerSettingLength)
+{
+    return AsCoreType(aInstance).Get<Utils::PowerCalibration>().GetRawPowerSetting(aChannel, aRawPowerSetting,
```

Comment:
Can consider doing the nun-null checks here instead of the `GetRawPowerSettings`?